### PR TITLE
add mx.trace (#1143)

### PR DESCRIPTION
**Pull request by abeleinin**
## Adds mx.trace support

Implements `mx.trace` (#1143) in C++ and Python.

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

File: python/src/ops.cpp [hunk 1]
```diff
@@ -4052,6 +4052,45 @@ void init_ops(nb::module_& m) {
...
+      },
+      nb::arg(),
+      "offset"_a = 0,
+      "axis1"_a = 0,
+      "axis2"_a = 1,
+      "dtype"_a = nb::none(),
+      nb::kw_only(),
+      "stream"_a = nb::none(),
+      nb::sig(
+          "def trace(a: array, offset: int = 0, axis1: int = 0, axis2: int = 1, dtype = Optional[Dtype] = None, stream: Union[None, Stream, Device] = None) -> array"),
```
Comment (awni) [inline python/src/ops.cpp:4076]:
```suggestion
          "def trace(a: array, /, offset: int = 0, axis1: int = 0, axis2: int = 1, dtype = Optional[Dtype] = None, *, stream: Union[None, Stream, Device] = None) -> array"),
```

**Review by awni — COMMENTED** [review]
This is really nicely done, thanks! I left two minor comments. Could you address then we can merge?

File: mlx/ops.cpp [hunk 2]
```diff
@@ -4114,6 +4114,59 @@ array diag(const array& a, int k /* = 0 */, StreamOrDevice s /* = {} */) {
...
+    throw std::out_of_range(msg.str());
+  }
+
+  if (ax1 == ax2) {
+    throw std::invalid_argument(
+        "[trace] axis1 and axis2 cannot be the same axis");
+  }
+
+  return astype(
+      sum(diagonal(a, offset, axis1, axis2), -1), dtype, to_stream(s));
```
Comment (awni) [inline mlx/ops.cpp:4154]: One more thing, I would do the `astype` before the `sum`. Part of the purpose of the `dtype` is to enable the reduction in higher precision if needed. So if you take the trace of an int8 matrix and you want the reduction in `int32` so it doesn't overflow.

**Review by awni — APPROVED** [review]
Thanks!!

File: mlx/ops.cpp [hunk 3]
```diff
@@ -4113,6 +4113,63 @@ array diag(const array& a, int k /* = 0 */, StreamOrDevice s /* = {} */) {
...
+    int offset,
+    int axis1,
+    int axis2,
+    StreamOrDevice s /* = {} */) {
+  auto dtype = a.dtype();
+  return trace(a, offset, axis1, axis2, dtype, to_stream(s));
+}
+array trace(const array& a, StreamOrDevice s /* = {} */) {
+  auto dtype = a.dtype();
+  return trace(a, 0, 0, 1, dtype, to_stream(s));
```
Comment (awni) [inline mlx/ops.cpp:4170]: Actually, one last thing. You don't need to call `to_stream` here every time. You should be able to simply pass in the stream as `s`. Will be a bit more readable.  Do you mind updating that?

Comment (abeleinin) [reply mlx/ops.cpp:4170]: sure thing!

**Pull request merged**